### PR TITLE
[improve][broker] Make MessageRedeliveryController work more efficiently

### PR DESCRIPTION
**Pull request by codelipenghui**
### Motivation

The implementation of MessageRedeliveryController is inefficient

- Duplicated Ledger IDs https://github.com/apache/pulsar/blob/master/pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentSortedLongPairSet.java#L50
- high sorting overhead even if you just want to get a few redeliveries https://github.com/apache/pulsar/pull/17804/files#diff-4c2d35368b21e5c33d65886196e8673c04a69f5c5cb888cdd04c29f8823502fdL109-L122

image

```
"BookKeeperClientWorker-OrderedExecutor-1-0" #40 prio=5 os_prio=0 cpu=27401742.31ms elapsed=788038.50s tid=0x00007f4cf95bc800 nid=0x56 runnable  [0x00007f4cd49a0000]
   java.lang.Thread.State: RUNNABLE
	at java.util.TreeMap.put(java.base@11.0.15/TreeMap.java:566)
	at java.util.TreeSet.add(java.base@11.0.15/TreeSet.java:255)
	at org.apache.pulsar.common.util.collections.ConcurrentSortedLongPairSet.lambda$items$4(ConcurrentSortedLongPairSet.java:152)
	at org.apache.pulsar.common.util.collections.ConcurrentSortedLongPairSet$$Lambda$1150/0x00000008408d9040.accept(Unknown Source)
	at org.apache.pulsar.common.util.collections.ConcurrentLongPairSet$Section.forEach(ConcurrentLongPairSet.java:563)
	at org.apache.pulsar.common.util.collections.ConcurrentLongPairSet.forEach(ConcurrentLongPairSet.java:242)
	at org.apache.pulsar.common.util.collections.ConcurrentSortedLongPairSet.items(ConcurrentSortedLongPairSet.java:151)
	at org.apache.pulsar.common.util.collections.ConcurrentSortedLongPairSet.items(ConcurrentSortedLongPairSet.java:143)
	at org.apache.pulsar.common.util.collections.ConcurrentSortedLongPairSet.items(ConcurrentSortedLongPairSet.java:128)
	at org.apache.pulsar.broker.service.persistent.MessageRedeliveryController.getMessagesToReplayNow(MessageRedeliveryController.java:112)
	at org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers.getMessagesToReplayNow(PersistentDispatcherMultipleConsumers.java:850)
	- eliminated <0x0000000757da7548> (a org.apache.pulsar.broker.service.persistent.PersistentStickyKeyDispatcherMultipleConsumers)
	at org.apache.pulsar.broker.service.persistent.PersistentStickyKeyDispatcherMultipleConsumers.getMessagesToReplayNow(PersistentStickyKeyDispatcherMultipleConsumers.java:432)
	- locked <0x0000000757da7548> (a org.apache.pulsar.broker.service.persistent.PersistentStickyKeyDispatcherMultipleConsumers)
	at org.apache.pulsar.broker.service.persistent.PersistentStickyKeyDispatcherMultipleConsumers.sendMessagesToConsumers(PersistentStickyKeyDispatcherMultipleConsumers.java:175)
	at org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers.readEntriesComplete(PersistentDispatcherMultipleConsumers.java:483)
	- locked <0x0000000757da7548> (a org.apache.pulsar.broker.service.persistent.PersistentStickyKeyDispatcherMultipleConsumers)
	at org.apache.bookkeeper.mledger.impl.ManagedCursorImpl$11.readEntryComplete(ManagedCursorImpl.java:1326)
	- locked <0x00000007129d8210> (a org.apache.bookkeeper.mledger.impl.ManagedCursorImpl$11)
	at org.apache.bookkeeper.mledger.impl.EntryCacheImpl.lambda$asyncReadEntry0$0(EntryCacheImpl.java:222)
	at org.apache.bookkeeper.mledger.impl.EntryCacheImpl$$Lambda$1158/0x00000008408db440.accept(Unknown Source)
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(java.base@11.0.15/CompletableFuture.java:714)
	at java.util.concurrent.CompletableFuture$Completion.run(java.base@11.0.15/CompletableFuture.java:478)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@11.0.15/ThreadPoolExecutor.java:1128)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@11.0.15/ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(java.base@11.0.15/Thread.java:829)
```

The issue usually happens if you have a large number of backlogs with the Key_Shared subscription. The publish latency will go high due to the IO thread and `BookKeeperClientWorker` thread with high high CPU usage

image

A related fix #15354, but the fix can't fix the case that the `MessageRedeliveryController` have many redelivery messages
This PR can make the `MessageRedeliveryController` work more efficiently by introducing a new Ordered Map and Bitmap based LongPair Ordered Set

-----------------
And the issue can be reproduced by standalone and pulsar-perf

1. Start a standalone
2. Use pulsar-perf to produce messages `bin/pulsar-perf produce test -r 1000 -bm 1 -mk random`
3. Open 3 terminals to run `bin/pulsar-perf consume -st Key_Shared -r 100 -n 50 test`
4. After a while, stop 2 perf consume process
5. You will see the publish latency increased

image

With this PR, the situation will not happen

### Modifications

- Added [RoaringBitmap](https://roaringbitmap.org/) dependency which is a widely used compressed bitset. The dependency is introduced to broker, not the common module because the common module will also expose the Java Client
- Added BitmapSortedLongPairSet for MessageRedeliveryController
- Update the MessageRedeliveryController to use BitmapSortedLongPairSet for messagesToRedeliver

### Verifying this change

New test added BitmapSortedLongPairSetTest

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc-required` 
(Your PR needs to update docs, and you will update later)

- [x] `doc-not-needed` 
(Please explain why)

- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### Matching PR in forked repository

PR in forked repository: https://github.com/codelipenghui/incubator-pulsar/pull/14

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/utils/BitmapSortedLongPairSet.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        try {
+            for (Map.Entry<Long, RoaringBitmap> entry : map.entrySet()) {
+                Iterator<Integer> iterator = entry.getValue().stream().iterator();
+                while (iterator.hasNext() && items.size() < numberOfItems) {
+                    items.add(longPairConverter.apply(entry.getKey(), iterator.next()));
+                }
+                if (items.size() == numberOfItems) {
+                    break;
+                }
+            }
```
Comment (315157973) [inline pulsar-broker/src/main/java/org/apache/pulsar/utils/BitmapSortedLongPairSet.java:110]: `map.subMap()`  Can it be changed to this?

**Review by merlimat — APPROVED** [review]
👍 Nice!

File: pulsar-broker/src/main/java/org/apache/pulsar/utils/BitmapSortedLongPairSet.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        lock.readLock().lock();
+        try {
+            RoaringBitmap bitSet = map.get(item1);
+            return bitSet != null && bitSet.contains(item2, item2 + 1);
+        } finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    public synchronized void removeUpTo(long item1, long item2) {
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/utils/BitmapSortedLongPairSet.java:73]: Question: why do we need `synchronized` here?

**Review by mattisonchao — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/utils/BitmapSortedLongPairSet.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+import java.util.NavigableSet;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.apache.pulsar.common.util.collections.LongPairSet;
+import org.roaringbitmap.RoaringBitmap;
+
+public class BitmapSortedLongPairSet {
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/utils/BitmapSortedLongPairSet.java:33]: Maybe we can indicate this class is thread-safe by adding a comment or name like `Concurrent***`?

File: pulsar-broker/src/main/java/org/apache/pulsar/utils/BitmapSortedLongPairSet.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    }
+
+    public synchronized void removeUpTo(long item1, long item2) {
+        lock.writeLock().lock();
+        try {
+            Map.Entry<Long, RoaringBitmap> firstEntry = map.firstEntry();
+            while (firstEntry != null && firstEntry.getKey() <= item1) {
+                if (firstEntry.getKey() < item1) {
+                    map.remove(firstEntry.getKey(), firstEntry.getValue());
```
Comment (315157973) [inline pulsar-broker/src/main/java/org/apache/pulsar/utils/BitmapSortedLongPairSet.java:79]: Is it possible to avoid loop traversal by using `subMap`?

Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/BitmapSortedLongPairSet.java:79]: Sub map will have the low and upper boundaries. If we change to sub map here e.g. from 3 to 10, you will not able to add 2 or 11 again.

File: pulsar-broker/src/main/java/org/apache/pulsar/utils/BitmapSortedLongPairSet.java (hunk 2 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/BitmapSortedLongPairSet.java:73]: 😁, I copied the method signature from another class

File: pulsar-broker/src/main/java/org/apache/pulsar/utils/ConcurrentBitmapSortedLongPairSet.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+public class ConcurrentBitmapSortedLongPairSet {
+
+    private final NavigableMap<Long, RoaringBitmap> map = new TreeMap<>();
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+    public void add(long item1, long item2) {
+        lock.writeLock().lock();
+        try {
+            RoaringBitmap bitSet = map.computeIfAbsent(item1, k -> new RoaringBitmap());
+            bitSet.add(item2, item2 + 1);
```
Comment (coderzc) [inline pulsar-broker/src/main/java/org/apache/pulsar/utils/ConcurrentBitmapSortedLongPairSet.java:42]: The `RoaringbitMap` can't support the number of more than`0xffffffff`, do we need to replace it with `Roaring64bitmap`?

Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/ConcurrentBitmapSortedLongPairSet.java:42]: @coderzc This doesn't actually happen in practice. And the individual acks of the ManagedCursor also can't support 64bit entry ID. Maybe we can consider supporting 64bit entry ID in the future if we really need it.

Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/utils/ConcurrentBitmapSortedLongPairSet.java:42]: Ok, I see.

**Review by coderzc — APPROVED** [review]

**Review by 315157973 — APPROVED** [review]

**Pull request merged**